### ISSUE TITLE
A question about parameter z_t

### ISSUE THREAD

**Issue by simpleoier** [facebookresearch/adaptive-span#15]
It's a good work.
I am so sorry to ask a stupid question.  From my understanding, the [current_val](https://github.com/facebookresearch/adaptive-span/blob/a8d90b8a8481ef1ae50a73b696c290aa88d34744/adaptive_span.py#L34) represents the span parameter z_t. Is this correct?

**Comment by prajjwal1** [comment]
`current_val` is responsible for determining the attention span. It is initialized with zero, and is updated as per `optimizer.step()` operation. Is the adaptive_span_loss updating in your experiments ?

**Comment by tesatory** [comment]
Not exactly. Variable current_val is equal to z divided by the maximum allowed span, so it always lie within [0,1].

**Comment by prajjwal1** [comment]
Thanks for clarifying.

**Comment by simpleoier** [comment]
Thanks for your answers.  
In this way, the attention span is a learnable parameter. But after training, the span is fixed, isn't it? It cannot change corresponding to different inputs x.
While in the [paper](https://arxiv.org/pdf/1905.07799.pdf), Sec 2.2, you mentioned other methods where the span parameter z_t is computed as z_t = S \sigma(v^T x_t + b). Why don't you use this kind of method to determine the span? Did you do comparisons against such methods?

**Comment by simpleoier** [comment]
> `current_val` is responsible for determining the attention span. It is initialized with zero, and is updated as per `optimizer.step()` operation. Is the adaptive_span_loss updating in your experiments ?

I am only reading the paper and the source code just because it is interesting. I haven't done experiments with this method. But if you have any interesting experiences or suggestions when you ran the codes, I would like to discuss it with you. Thank you very much.

**Comment by simpleoier** [comment]
> Thanks for your answers.
> In this way, the attention span is a learnable parameter. But after training, the span is fixed, isn't it? It cannot change corresponding to different inputs x.
> While in the [paper](https://arxiv.org/pdf/1905.07799.pdf), Sec 2.2, you mentioned other methods where the span parameter z_t is computed as z_t = S \sigma(v^T x_t + b). Why don't you use this kind of method to determine the span? Did you do comparisons against such methods?

@tesatory I think I know it. The method in my question is the dynamic span. And the comparison is in the Table 3. For adaptive span, the span is fixed after training.  Is it correct?

**Comment by prajjwal1** [comment]
> > `current_val` is responsible for determining the attention span. It is initialized with zero, and is updated as per `optimizer.step()` operation. Is the adaptive_span_loss updating in your experiments ?
> 
> @prajjwal1 I am only reading the paper and the source code just because it is interesting. I haven't done experiments with this method. But if you have any interesting experiences or suggestions when you ran the codes, I would like to discuss it with you. Thank you very much.

Thanks. I've integrated this approach with my work. After some struggling, I finally for the `current_val` to work as intended. But my accuracy is not comparable to standard attention. I think the mask (which is being generated from `current_val` and `mask_template` is harming the performance rather than helping it as per my experiments.

**Comment by prajjwal1** [comment]
> > Thanks for your answers.
> > In this way, the attention span is a learnable parameter. But after training, the span is fixed, isn't it? It cannot change corresponding to different inputs x.
> > While in the [paper](https://arxiv.org/pdf/1905.07799.pdf), Sec 2.2, you mentioned other methods where the span parameter z_t is computed as z_t = S \sigma(v^T x_t + b). Why don't you use this kind of method to determine the span? Did you do comparisons against such methods?
> 
> @tesatory I think I know it. The method in my question is the dynamic span. And the comparison is in the Table 3. For adaptive span, the span is fixed after training. Is it correct?

I've some doubts regarding inference time. How will span change when the weights are kept constant ? This is something I need to work on next.

**Comment by tesatory** [comment]
> > Thanks for your answers.
> > In this way, the attention span is a learnable parameter. But after training, the span is fixed, isn't it? It cannot change corresponding to different inputs x.
> > While in the [paper](https://arxiv.org/pdf/1905.07799.pdf), Sec 2.2, you mentioned other methods where the span parameter z_t is computed as z_t = S \sigma(v^T x_t + b). Why don't you use this kind of method to determine the span? Did you do comparisons against such methods?
> 
> @tesatory I think I know it. The method in my question is the dynamic span. And the comparison is in the Table 3. For adaptive span, the span is fixed after training. Is it correct?

Yes, the span is fixed once training is finished. This is not the case for the dynamic-span, which we didn't include in the code for simplicity. Also, there was a little practical advantage with the dynamic-span. But you can easily implement it if you want.

**Comment by ttambe** [comment]
@tesatory, would it be possible to share the dynamic-span code?
